### PR TITLE
Frontend+Backend: Add feature to close participation in the challenge even when it is ongoing

### PR DESCRIPTION
**Pull request by Sanji515**
- Added a switch for toggling the participation
- Added a new boolean field `can_participate` in challenge model
- Used existing API `challenge_detail` for updating the boolean variable

Here's the screenshot:

Screenshot from 2019-05-31 06-39-29

Screenshot from 2019-05-31 06-39-38

### CODE REVIEW

File: apps/challenges/models.py [hunk 1]
```diff
@@ -56,6 +56,7 @@ def __init__(self, *args, **kwargs):
     published = models.BooleanField(
         default=False, verbose_name="Publicly Available", db_index=True
     )
+    can_participate = models.BooleanField(default=True)
```
Comment (RishabhJain2018) [inline apps/challenges/models.py]: Can we please rename`can_participate` to `is_registration_open`?

Comment (Sanji515) [reply apps/challenges/models.py]: @RishabhJain2018 done

File: frontend/src/views/web/challenge/challenge-page.html [hunk 2]
```diff
@@ -48,14 +48,22 @@
             <div class="col m3 l3 align-right">
                 <button class="waves-effect waves-dark btn ev-btn-dark w-300 fs-14" type="submit" ng-click="challenge.starChallenge()" ng-init="challenge.isStarred()"><i class="fa fa-star star"></i> {{challenge.data}}&nbsp;&nbsp;{{challenge.count}}</button>
             </div>
+            <div ng-if="challenge.isChallengeHost" style="padding-top: 30px;" class="col m3 l3 align-right">
+                <div style="display: inline-block;" ng-click="challenge.toggleParticipation($event, challenge.canParticipate)"
+                    class="btn-switch" ng-class="{'btn-switch--on':challenge.canParticipate}" ng-model="challenge.canParticipate">
```
Comment (RishabhJain2018) [inline frontend/src/views/web/challenge/challenge-page.html]: I think you missed updating this - `challenge.canParticipate`

Comment (RishabhJain2018) [reply frontend/src/views/web/challenge/challenge-page.html]: Please update it everywhere.

**Comment by RishabhJain2018** [comment]
Also, please add a screenshot of it.

File: frontend/src/views/web/challenge/challenge-page.html (hunk 2 above)
Comment (Sanji515) [reply frontend/src/views/web/challenge/challenge-page.html]: yeah I noticed that after pushing :)

**Comment by RishabhJain2018** [comment]
Also, please the check in this API as well https://github.com/Cloud-CV/EvalAI/blob/master/apps/challenges/views.py#L227

You might need to add/fix the tests as well.

File: frontend/src/views/web/challenge/challenge-page.html (hunk 2 above)
Comment (Sanji515) [reply frontend/src/views/web/challenge/challenge-page.html]: @RishabhJain2018 done

**Comment by Sanji515** [comment]
@RishabhJain2018 

> Also, please add a screenshot of it

done

> You might need to add/fix the tests as well

I've already fixed that test, please let me know if something more needed

File: apps/challenges/views.py [hunk 3]
```diff
@@ -233,6 +233,12 @@ def add_participant_team_to_challenge(
     except Challenge.DoesNotExist:
         response_data = {"error": "Challenge does not exist"}
         return Response(response_data, status=status.HTTP_406_NOT_ACCEPTABLE)
+    
+    if not challenge.is_registration_open:
```
Comment (RishabhJain2018) [inline apps/challenges/views.py]: Please add tests for this condition in the API.

Comment (Sanji515) [reply apps/challenges/views.py]: @RishabhJain2018 done

File: frontend/src/js/controllers/challengeCtrl.js [hunk 4]
```diff
@@ -269,6 +270,48 @@
...
+        vm.toggleParticipation = function (ev, isRegistrationOpen) {
+            // ev.stopPropagation();
+            var participationState;
+            if (isRegistrationOpen) {
+                participationState = 'close';
+            } else {
+                participationState = 'open';
+            }
+            var confirm = $mdDialog.confirm()
+                          .title('You really want to ' + participationState + ' the participation?')
```
Comment (RishabhJain2018) [inline frontend/src/js/controllers/challengeCtrl.js]: `'You really want to ' + participationState + ' the participation?'`
 -> `'participationState + ' participation in the challenge?'`

File: frontend/src/js/controllers/challengeCtrl.js [hunk 5]
```diff
@@ -269,6 +270,48 @@
 
         utilities.sendRequest(parameters);
 
+        vm.toggleParticipation = function (ev, isRegistrationOpen) {
+            // ev.stopPropagation();
+            var participationState;
+            if (isRegistrationOpen) {
+                participationState = 'close';
+            } else {
+                participationState = 'open';
```
Comment (RishabhJain2018) [inline frontend/src/js/controllers/challengeCtrl.js]: `open` -> `Open`

File: frontend/src/js/controllers/challengeCtrl.js [hunk 6]
```diff
@@ -269,6 +270,48 @@
 
         utilities.sendRequest(parameters);
 
+        vm.toggleParticipation = function (ev, isRegistrationOpen) {
+            // ev.stopPropagation();
+            var participationState;
+            if (isRegistrationOpen) {
+                participationState = 'close';
```
Comment (RishabhJain2018) [inline frontend/src/js/controllers/challengeCtrl.js]: `close` -> `Close`

File: frontend/src/js/controllers/challengeCtrl.js [hunk 7]
```diff
@@ -269,6 +270,48 @@
...
+            } else {
+                participationState = 'open';
+            }
+            var confirm = $mdDialog.confirm()
+                          .title('You really want to ' + participationState + ' the participation?')
+                          .ariaLabel('')
+                          .targetEvent(ev)
+                          .ok('I\'m sure')
```
Comment (RishabhJain2018) [inline frontend/src/js/controllers/challengeCtrl.js]: `Yes, I\'m sure`

File: frontend/src/js/controllers/challengeCtrl.js [hunk 8]
```diff
@@ -269,6 +270,48 @@
...
+                          .cancel('No.');
```
Comment (RishabhJain2018) [inline frontend/src/js/controllers/challengeCtrl.js]: `No`

File: frontend/src/js/controllers/challengeCtrl.js (hunk 4 above)
Comment (Sanji515) [reply frontend/src/js/controllers/challengeCtrl.js]: done

File: frontend/src/js/controllers/challengeCtrl.js (hunk 5 above)
Comment (Sanji515) [reply frontend/src/js/controllers/challengeCtrl.js]: done

File: frontend/src/js/controllers/challengeCtrl.js (hunk 6 above)
Comment (Sanji515) [reply frontend/src/js/controllers/challengeCtrl.js]: done

File: frontend/src/js/controllers/challengeCtrl.js (hunk 7 above)
Comment (Sanji515) [reply frontend/src/js/controllers/challengeCtrl.js]: done

File: frontend/src/js/controllers/challengeCtrl.js (hunk 8 above)
Comment (Sanji515) [reply frontend/src/js/controllers/challengeCtrl.js]: done

**Pull request merged**